federation/pkg/federation-controller/replicaset/planner/planer.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package planer
+
+import (
+	"sort"
+
+	fed_api "k8s.io/kubernetes/federation/apis/federation"
+)
+
+// Planer decides how many out of the given replicas should be placed in each of the
+// federated clusters.
+type Planer struct {
+	preferences *fed_api.FederatedReplicaSetPreferences
+}
+
+type namedLocalReplicaSetPreferences struct {
+	clusterName string
+	fed_api.LocalReplicaSetPreferences
+}
+
+type byWeight []*namedLocalReplicaSetPreferences
+
+func (a byWeight) Len() int      { return len(a) }
+func (a byWeight) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// Preferences are sorted according by decreasing weight and increasing clusterName.
+func (a byWeight) Less(i, j int) bool {
+	return (a[i].Weight > a[j].Weight) || (a[i].Weight == a[j].Weight && a[i].clusterName < a[j].clusterName)
+}
+
+func NewPlaner(preferences *fed_api.FederatedReplicaSetPreferences) *Planer {
+	return &Planer{
+		preferences: preferences,
+	}
+}
+
+// Distribute the desired number of replicas among the given cluster according to the planner preferences.
+func (p *Planer) Plan(replicasToDistribute int64, availableClusters []string) map[string]int64 {
+	preferences := make([]*namedLocalReplicaSetPreferences, 0, len(availableClusters))
+
+	named := func(name string, pref fed_api.LocalReplicaSetPreferences) *namedLocalReplicaSetPreferences {
+		return &namedLocalReplicaSetPreferences{
+			clusterName:                name,
+			LocalReplicaSetPreferences: pref,
+		}
+	}
+
+	for _, cluster := range availableClusters {
+		if localRSP, found := p.preferences.Clusters[cluster]; found {
+			preferences = append(preferences, named(cluster, localRSP))
+		} else {
+			if localRSP, found := p.preferences.Clusters["*"]; found {
+				preferences = append(preferences, named(cluster, localRSP))
+			} else {
+				preferences = append(preferences, named(cluster, fed_api.LocalReplicaSetPreferences{Weight: 0}))
```

Comment:
nit: Do we really need an explicit weight of zero.  Can't we just leave the cluster out of the list of eligible clusters?